core/src/main/java/io/questdb/griffin/engine/functions/groupby/StringAggGroupByFunctionFactory.java

```diff
@@ -0,0 +1,120 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2020 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.groupby;
+
+import io.questdb.cairo.ArrayColumnTypes;
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.ColumnType;
+import io.questdb.cairo.map.MapValue;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.GroupByFunction;
+import io.questdb.griffin.engine.functions.StrFunction;
+import io.questdb.std.ObjList;
+
+public class StringAggGroupByFunctionFactory implements FunctionFactory {
+
+    @Override
+    public String getSignature() {
+        return "string_agg(Sa)";
+    }
+
+    @Override
+    public boolean isGroupBy() {
+        return true;
+    }
+
+    @Override
+    public Function newInstance(ObjList<Function> args, int position, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) {
+        return new StringAggGroupByFunction(position, args.getQuick(0), args.getQuick(1).getChar(null));
+    }
+
+    private static class StringAggGroupByFunction extends StrFunction implements GroupByFunction {
+        private final Function arg;
+        private final char delimiter;
+        private int valueIndex;
+        private boolean empty;
+
+        public StringAggGroupByFunction(int position, Function arg, char delimiter) {
+            super(position);
+            this.arg = arg;
+            this.delimiter = delimiter;
+        }
+
+        @Override
+        public void computeFirst(MapValue mapValue, Record record) {
+            CharSequence str = arg.getStr(record);
```

Comment:
I would like to suggest to implement this function slightly differently:

- StringAggFunction can maintain dense list of DirectCharSink instances
- the map supplied into aggregate function can be used to key the indexes in the above list

The reason for suggesting new approach is that existing approach will not work for keyed aggregations, it produces string instances on value read and adds narrow-use-case functionality to fairly generic `SimpleMapValue`

Reply:
Great suggestion, thanks! 

In fact, there is no need to maintain a list of DirectCharSink because each instance of StringAggGroupByFunction is linked with only one column. 

When it comes to the initial capacity of the sink, for now, I’ve decided to hardcode the value, without adding a new property to the configuration. I chose the value based on the results of the benchmark with the query from pgjdbc. Perhaps in the future, if this functionality will be used more widely, it will be worth making this value configurable. What do you think?

Reply:
Constant capacity is fine for this case. Also I saw you have a test with keyed aggregation (unsupported check). Its fine to have single sink per function.